From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Nora Schiffer <nora.schiffer@ew.tq-group.com>
Date: Wed, 29 May 2024 09:23:12 +0200
Subject: [PATCH] gpio: tqmx86: prepare support for variants with more GPIO
 channels

The SMARC variants of TQMx86 modules have more than 8 GPIOs and use a
different mapping between GPIOs channels and IRQ register bits. Prepare
their support by turning a number of constants into tqmx86_gpio_data
fields.

No functional change intended.

Signed-off-by: Nora Schiffer <nora.schiffer@ew.tq-group.com>
Upstream-Status: Pending
---
 drivers/gpio/gpio-tqmx86.c | 35 +++++++++++++++++++----------------
 1 file changed, 19 insertions(+), 16 deletions(-)

diff --git a/drivers/gpio/gpio-tqmx86.c b/drivers/gpio/gpio-tqmx86.c
index 27dd09273292a8db1b78c1f98aaa950b52d1e9be..a4a0623cd18f86ff786b50839fed43b82c20c92b 100644
--- a/drivers/gpio/gpio-tqmx86.c
+++ b/drivers/gpio/gpio-tqmx86.c
@@ -14,14 +14,12 @@
 #include <linux/interrupt.h>
 #include <linux/kernel.h>
 #include <linux/module.h>
+#include <linux/platform_data/gpio-tqmx86.h>
 #include <linux/platform_device.h>
 #include <linux/pm_runtime.h>
 #include <linux/seq_file.h>
 #include <linux/slab.h>
 
-#define TQMX86_NGPIO	8
-#define TQMX86_NGPO	4	/* 0-3 - output */
-#define TQMX86_NGPI	4	/* 4-7 - input */
 #define TQMX86_DIR_INPUT_MASK	0xf0	/* 0-3 - output, 4-7 - input */
 
 #define TQMX86_GPIODD	0	/* GPIO Data Direction Register */
@@ -49,10 +47,13 @@ struct tqmx86_gpio_data {
 	struct gpio_chip	chip;
 	void __iomem		*io_base;
 	int			irq;
+	unsigned int		irq_offset;
+	unsigned int		nirq;
+	DECLARE_BITMAP(irq_valid_mask, TQMX86_NGPIO_MAX);
 	/* Lock must be held for accessing output and irq_type fields */
 	raw_spinlock_t		spinlock;
-	DECLARE_BITMAP(output, TQMX86_NGPIO);
-	u8			irq_type[TQMX86_NGPIO];
+	DECLARE_BITMAP(output, TQMX86_NGPIO_MAX);
+	u8			irq_type[TQMX86_NGPIO_MAX];
 };
 
 static u8 tqmx86_gpio_read(struct tqmx86_gpio_data *gd, unsigned int reg)
@@ -148,8 +149,8 @@ static int tqmx86_gpio_get_direction(struct gpio_chip *chip,
 static void tqmx86_gpio_irq_config(struct tqmx86_gpio_data *gpio, int hwirq)
 	__must_hold(&gpio->spinlock)
 {
+	int gpiic_irq = hwirq - gpio->irq_offset;
 	u8 type = TQMX86_INT_TRIG_NONE;
-	int gpiic_irq = hwirq - TQMX86_NGPO;
 
 	if (gpio->irq_type[hwirq] & TQMX86_INT_UNMASKED) {
 		type = gpio->irq_type[hwirq] & TQMX86_INT_TRIG_MASK;
@@ -239,8 +240,8 @@ static void tqmx86_gpio_irq_handler(struct irq_desc *desc)
 	irq_bits = irq_status;
 
 	scoped_guard(raw_spinlock_irqsave, &gpio->spinlock) {
-		for_each_set_bit(i, &irq_bits, TQMX86_NGPI) {
-			hwirq = i + TQMX86_NGPO;
+		for_each_set_bit(i, &irq_bits, gpio->nirq) {
+			hwirq = i + gpio->irq_offset;
 
 			/*
 			 * Edge-both triggers are implemented by flipping the
@@ -271,9 +272,9 @@ static void tqmx86_gpio_irq_handler(struct irq_desc *desc)
 		}
 	}
 
-	for_each_set_bit(i, &irq_bits, TQMX86_NGPI)
+	for_each_set_bit(i, &irq_bits, gpio->nirq)
 		generic_handle_domain_irq(gpio->chip.irq.domain,
-					  i + TQMX86_NGPO);
+					  i + gpio->irq_offset);
 
 	chained_irq_exit(irq_chip, desc);
 }
@@ -298,11 +299,9 @@ static void tqmx86_init_irq_valid_mask(struct gpio_chip *chip,
 				       unsigned long *valid_mask,
 				       unsigned int ngpios)
 {
-	/* Only GPIOs 4-7 are valid for interrupts. Clear the others */
-	clear_bit(0, valid_mask);
-	clear_bit(1, valid_mask);
-	clear_bit(2, valid_mask);
-	clear_bit(3, valid_mask);
+	struct tqmx86_gpio_data *gpio = gpiochip_get_data(chip);
+
+	bitmap_and(valid_mask, valid_mask, gpio->irq_valid_mask, ngpios);
 }
 
 static void tqmx86_gpio_irq_print_chip(struct irq_data *d, struct seq_file *p)
@@ -351,6 +350,8 @@ static int tqmx86_gpio_probe(struct platform_device *pdev)
 
 	raw_spin_lock_init(&gpio->spinlock);
 	gpio->io_base = io_base;
+	gpio->irq_valid_mask[0] = TQMX86_DIR_INPUT_MASK;
+	gpio->irq_offset = 4;
 
 	tqmx86_gpio_write(gpio, (u8)~TQMX86_DIR_INPUT_MASK, TQMX86_GPIODD);
 
@@ -371,7 +372,7 @@ static int tqmx86_gpio_probe(struct platform_device *pdev)
 	chip->get_direction = tqmx86_gpio_get_direction;
 	chip->get = tqmx86_gpio_get;
 	chip->set = tqmx86_gpio_set;
-	chip->ngpio = TQMX86_NGPIO;
+	chip->ngpio = 8;
 	chip->parent = pdev->dev.parent;
 
 	pm_runtime_enable(&pdev->dev);
@@ -379,6 +380,8 @@ static int tqmx86_gpio_probe(struct platform_device *pdev)
 	if (irq > 0) {
 		u8 irq_status;
 
+		gpio->nirq = 4;
+
 		/* Mask all interrupts */
 		tqmx86_gpio_write(gpio, 0, TQMX86_GPIIC);
 
